### ISSUE TITLE
Add documentation on parallel solves of a JuMP model

### ISSUE THREAD

**Issue by odow** [jump-dev/JuMP.jl#2348]
I get this question quite frequently. A section should be added to the docs explaining the different options.

Here is a great answer from SO: https://stackoverflow.com/questions/64177415/create-jump-model-with-multi-threading-in-julia

We should add a section on using `Distributed`. Here is an old answer from Discourse: https://discourse.julialang.org/t/parallel-solves-in-gurobi-jl/9908/8?u=odow

**Comment by odow** [comment]
@jd-lara asks: Add a tutorial on how to get the "duals" of a MIP (i.e., solve, fix the integers, convert to LP, and re-solve). 

We had some code to do so in PowerSimulations (and gitter), but writing a nice tutorial would be good.

**Comment by odow** [comment]
We should also add in-place solves for the subproblems of the column generation and Benders decomposition tutorials: 
https://jump.dev/JuMP.jl/dev/tutorials/algorithms/benders_decomposition/

**Comment by odow** [comment]
- [ ] add a tutorial (e.g., AC-OPF) using complex number support: https://github.com/jump-dev/JuMP.jl/issues/3114.

**Comment by odow** [comment]
- [ ] add an introduction to optimization tutorial. See the discussion in https://github.com/jump-dev/JuMP.jl/issues/2859

**Comment by odow** [comment]
I added a bunch of examples for constraint programming: https://jump.dev/JuMP.jl/dev/tutorials/linear/constraint_programming/

We should do the same for conic. A lot of those sets aren't used very much because they're not easily visible from JuMP.

**Comment by odow** [comment]
We're at a point where I've done most of the documentation changes I wanted to do. We still need to add a tutorial/example on the complex-valued stuff, and potentially a general introduction to optimization.

I'm not sure if the JuMP docs are the right place for a general introduction. Does anyone have a list of teaching materials that we could steer people towards instead?

**Comment by jd-foster** [comment]
I think [Linear Programming with MATLAB](https://epubs.siam.org/doi/book/10.1137/1.9780898718775) is brilliant and only wish (1) it was open and (2) it had a Julia version. But that doesn't answer your question.

**Comment by odow** [comment]
- [ ] Multiple people have been confused about setting warm starts for Ipopt because they follow the conic tutorial: 
      -  https://stackoverflow.com/questions/75543689/primal-and-dual-warm-starts-jump-error-message
      - https://discourse.julialang.org/t/jump-model-warm-start-using-ipopt/92660

**Comment by blegat** [comment]
- [ ] quantum information, state discrimination suggested by @araujoms in https://discourse.julialang.org/t/error-when-constraining-variable-to-be-hermitian-psd/95223

**Comment by odow** [comment]
https://github.com/jump-dev/JuMP.jl/pull/3250 would solve the quantum information and complex-number tutorials.

**Comment by blegat** [comment]
- [ ] A tutorial playing with adding bridges, removing bridges and printing active bridges. Probably an examples with RootDet -> GeoMean -> (3 possible bridges for GeoMean) -> SCS would be good. It could be added at the end of https://jump.dev/JuMP.jl/stable/tutorials/conic/ellipse_approx/

**Comment by jd-foster** [comment]
Nested optimization problems seem to be coming up frequently. A tutorial on solving the inner problem, returning function and gradients and registering these? Already there are a few examples on discourse.

**Comment by odow** [comment]
> Nested optimization problems seem to be coming up frequently

We have https://jump.dev/JuMP.jl/stable/tutorials/nonlinear/user_defined_hessians/#Bilevel-optimization, but perhaps it needs to be it's own tutorial, rather than a section?

**Comment by odow** [comment]
This question comes up surprisingly often: 

https://discourse.julialang.org/t/connecting-a-simple-first-order-solver-to-solve-standard-form-linear-program-to-jump/95694

I can see it being a good feature in `Utilities`, or as a tutorial with Simplex? It'd also make a good exercise for teaching.

**Comment by jd-foster** [comment]
> We have https://jump.dev/JuMP.jl/stable/tutorials/nonlinear/user_defined_hessians/#Bilevel-optimization, but perhaps it needs to be it's own tutorial, rather than a section?

See #3274 . Might be too ruthless on Rosenbrock.

**Comment by blegat** [comment]
> This question comes up surprisingly often:
> 
> https://discourse.julialang.org/t/connecting-a-simple-first-order-solver-to-solve-standard-form-linear-program-to-jump/95694
> 
> I can see it being a good feature in `Utilities`, or as a tutorial with Simplex? It'd also make a good exercise for teaching.

I agree, I am also planning to convert this one into a JuMP tutorial https://github.com/blegat/SimpleConicADMM. This could even do SDP if we add scaled PSD back to MOI

**Comment by odow** [comment]
So. I thought I'd give this Chat GPT thing a go. I could actually be useful in refining parts of the docs?(!?) Per their policies, I think we're okay to use the suggested text:

<img width="687" alt="image" src="https://user-images.githubusercontent.com/8177701/226080844-83cb7e11-6c5f-4933-87a8-3a745a181694.png">

---

Oscar Dowson
I'm writing technical documentation for JuMP, an algebraic modeling language for mathematical programming built in the Julia programing language. 

Suggest an improvement to the following text:

[here I copied should_i_use.md]

--- 

The text provides a good overview of the key features of JuMP and the benefits of using it. Here are a few suggestions to improve the clarity and readability of the text:

Clarify the target audience: The text should clarify the target audience for JuMP. Who would benefit from using JuMP, and why? For example, is JuMP more suitable for researchers, students, or industry practitioners?
Define mathematical programming: The text assumes that the reader is familiar with mathematical programming. It may be helpful to provide a brief definition or explanation of mathematical programming for readers who are not familiar with it.
Expand on the constraints: The text mentions that JuMP can handle "a set of constraints", but does not provide any further details on what types of constraints JuMP can handle. It may be helpful to provide examples of the types of constraints that JuMP can handle.
Provide more information on solver-independent interface: The text mentions that JuMP uses a solver-independent interface, but does not provide much detail on what this means or why it is important. It may be helpful to explain what a solver-independent interface is, and why it makes it easy to change between different optimization software packages.
Clarify ease of embedding: The text mentions that JuMP is easy to embed in other workflows, but does not provide much detail on how this works. It may be helpful to provide examples of how JuMP can be embedded in different workflows, and why this is beneficial.
Provide more information on MPL license: The text mentions that JuMP is MPL licensed, but does not provide much detail on what this means or why it is important. It may be helpful to explain what the MPL license is, and why it allows JuMP to be embedded in commercial software.
Provide more information on advanced algorithmic techniques: The text mentions that JuMP provides access to advanced algorithmic techniques, but does not provide much detail on what these techniques are or why they are important. It may be helpful to provide examples of the types of techniques that JuMP supports, and why they are beneficial.
Overall, the text provides a good starting point for describing JuMP's features and benefits, but could benefit from more detail and clarification in some areas.

---
Oscar Dowson
Can you rewrite it for me?

---
Sure, here's a revised version of the text with the suggested improvements:

Should you use JuMP?

JuMP is an algebraic modeling language for mathematical optimization written in the Julia language. If you have a constrained optimization problem that can be formulated using the language of mathematical programming, JuMP may be a good fit for you.

When should you use JuMP?

JuMP is ideal for researchers, students, or industry practitioners who need to solve optimization problems that involve decision variables, a scalar- or vector-valued objective function, and a set of constraints. JuMP offers a number of benefits that make it an attractive choice for mathematical optimization:

User-friendly: JuMP's syntax mimics natural mathematical expressions, making it easier to write and understand optimization models.
Solver independence: JuMP uses a solver-independent interface provided by the MathOptInterface package, which makes it easy to switch between open-source and commercial optimization software packages ("solvers").
Ease of embedding: JuMP is written purely in Julia and provides automatic installation of many open-source solvers. This makes it easy to embed JuMP in other workflows, such as simulations, web servers, or decomposition algorithms. JuMP's syntax is constrained by Julia's syntax and functionality, but this trade-off is offset by the ease of embedding.
Speed: Benchmarking has shown that JuMP can create problems at similar speeds to special-purpose modeling languages such as AMPL. JuMP communicates with most solvers in memory, avoiding the need to write intermediary files.
Access to advanced algorithmic techniques: JuMP supports efficient in-memory re-solves of linear programs and provides access to solver-independent and solver-dependent callbacks. These features provide access to advanced algorithmic techniques that were previously only available through solver-specific or low-level C++ libraries.
When should you not use JuMP?

JuMP may not be the best fit for optimization problems that cannot be formulated as a mathematical program or problems that require specialized features that are not currently supported by JuMP or its solvers. If you need to use a specialized solver that is not currently supported by JuMP, you may need to look for a different optimization tool.

Overall, JuMP provides a powerful and flexible platform for solving mathematical optimization problems, and its user-friendly syntax, solver independence, ease of embedding, speed, and access to advanced algorithmic techniques make it an attractive choice for researchers, students, and industry practitioners alike.


I did something similar with index.md

<img width="788" alt="image" src="https://user-images.githubusercontent.com/8177701/226080781-cdbf69bc-50e0-45cd-92be-0327a610c33b.png">

**Comment by odow** [comment]
<img width="687" alt="image" src="https://user-images.githubusercontent.com/8177701/226081274-7ffeb863-55e9-4d3d-8ea2-cf775a9a73f0.png">

<img width="687" alt="image" src="https://user-images.githubusercontent.com/8177701/226081279-494a997a-6371-4841-9636-61cc2186d02d.png">

**Comment by odow** [comment]
Hi @francis-gagnon, sorry for the delay. I didn't notice this notification.

> It seems that the nested optimization tutorial is no longer working

That likely means you've installed an old version of JuMP. The docs are built and tested with every commit of JuMP, so things definitely work with the latest version.

> Also, I'm not sure that my understanding of the problem is good enough but, from the code, it seems that the first equation:

Yes, you're correct. I'll make the change.

**Comment by blegat** [comment]
Tutorial on using `Dualization.dual_optimizer` and update conic tutorials to make sure they don't use pure standard conic form with geometric conic form solver or vice versa:

- [ ] Dualize SCS for Luvász https://github.com/jump-dev/JuMP.jl/pull/3399/

**Comment by odow** [comment]
Working on dualization now.

**Comment by jd-foster** [comment]
Add basic dualization comparison to simple examples (#3408) : revisit if we get any comments or questions on discourse.

**Comment by odow** [comment]
@mlubin @jd-foster and I discussed a few documentation-related things at JuMP-dev. We could:

 * Add a video walkthrough for each tutorial
   * We could run each tutorial as a notebook with narration, like we would do in a tutorial
   * We could record a short (60 sec) snippet pointing out the key features
 * Develop course material for an introductory course in optimization
 * Refine the current set of tutorials into a textbook
   * We would need to resolve authorship and copyright questions

**Comment by odow** [comment]
The GAMS summation thing came up again, so we definitely need better docs:
https://discourse.julialang.org/t/creating-variables-from-a-vector-of-tuples-and-naming-them/103525

**Comment by odow** [comment]
Semi-integer and semi-continuous variables under-appreciated. As a concept, these are not well known, and I keep seeing people recommend the binary reformulation. I wonder what we could do to make these more prominent in the documentation.

**Comment by odow** [comment]
Add section on function tracing https://discourse.julialang.org/t/jump-user-defined-functions-with-vector-inputs-and-outputs/106025

**Comment by odow** [comment]
We should standardize the look of all plots in the documentation: https://github.com/jump-dev/JuMP.jl/pull/3569#issuecomment-1811814746

**Comment by jd-foster** [comment]
What do you think about switching to `PlotlyJS` for all the plots, get embedded interactivity, and drop the Plots dependency?

**Comment by jd-foster** [comment]
1. Can we have an Integer Programming section in the docs?
2. Is there an "Improved MIP modeling tricks" section i.e. [Mosek MIO](https://docs.mosek.com/modeling-cookbook/mio.html) but in JuMP?

**Comment by odow** [comment]
> What do you think about switching to PlotlyJS for all the plots, get embedded interactivity, and drop the Plots dependency?

Sure, provided it works with the PDF build? (I have a soft spot for Plotly https://dairyanalytics.co.nz)

**Comment by odow** [comment]
Yeah the linear section is getting a bit big. Okay to subdivide

**Comment by odow** [comment]
@blegat suggests adding an example using a third-party AD package: https://github.com/jump-dev/JuMP.jl/pull/3577#issuecomment-1821691334

**Comment by blegat** [comment]
We could write `plotlyjs()` to have a the plotly backend but I would stay with the Plots interface. Most people know this interface and now the plotly one

**Comment by odow** [comment]
We could add an MPC tutorial: https://github.com/jump-dev/JuMPTutorials.jl/pull/49

**Comment by odow** [comment]
Two of the top-10 discourse questions on jump.dev/forum are about the `sum` syntax:

 * https://discourse.julialang.org/t/sum-syntax/14589
 * https://discourse.julialang.org/t/how-to-use-the-summation-in-julia/41522

This suggests that people are searching for it. We should improve.

**Comment by odow** [comment]
Add tutorial on accessing basis: https://discourse.julialang.org/t/basis-and-non-basis-matrix/110219

**Comment by odow** [comment]
We need to add some "transitioning from XXX" tutorials. 

We keep getting questions from AMPL/GAMS/MATLAB users.

See this interesting discussion:
https://discourse.julialang.org/t/nonlinear-optimization-with-many-constraints-autodifferentiation-which-julia-solution/110678

**Comment by araujoms** [comment]
I can write a "transitioning from MATLAB" tutorial if you're interested.

**Comment by odow** [comment]
Sure! I don't really know what is needed or would be useful, so a draft would be super handy.

**Comment by araujoms** [comment]
Ok! I'll write what would have been useful for me as I transitioned from MATLAB to Julia, and hope it will also be useful for others.

**Comment by blegat** [comment]
I've had a request for a tutorial for getting the matrix form, e.g., doing this:
```julia
cache = SCS.OptimizerCache{Int}()
MOI.copy_to(cache, backend(model).optimizer.model.model_cache)
A = convert(SparseMatrixCSC{Float64,Int}, cache.constraints.coefficients)
b = cache.constraints.constants.b
```

**Comment by jd-foster** [comment]
With the new Transitioning section, it would be good to have a short referring section from the "[Should you use JuMP?](https://jump.dev/JuMP.jl/dev/should_i_use/#Should-you-use-JuMP?)" page
Also, I should write a "Transitioning from GAMS" one.

**Comment by jd-foster** [comment]
https://github.com/jump-dev/JuMP.jl/issues/3729#issuecomment-2062523491 should be a page in the docs.

**Comment by blegat** [comment]
Agreed, maybe it should mention `@profview_allocs` as well (by the way, I find this macro suprisingly difficult to find by googling, it should be advertised)

**Comment by odow** [comment]
On the documenter call, we discussed how to change logos in a GitHub readme based on light/dark-mode:
 * https://www.stefanjudis.com/notes/how-to-define-dark-light-mode-images-in-github-markdown/
 * https://github.com/Ferrite-FEM/Ferrite.jl/blob/aea4b5155e225352e483dbb07acf22fc28b430ac/README.md?plain=1#L1-L5

**Comment by odow** [comment]
We could add feasibility cuts to the Benders tutorial

x-ref https://discourse.julialang.org/t/dual-ray-from-lp-solved-by-dual-algorithm/114068/4?u=odow

**Comment by joaquimg** [comment]
I'd say that should be either a second part of the tutorial or even a separate tutorial.
Adding all of it might make the simple Benders unnecessarily complex.

**Comment by blegat** [comment]
We can actually also add unbounded cuts (it's in https://github.com/JuliaStochOpt/StructDualDynProg.jl). The idea is as follows: if the parent is unbounded, you compute the unbounded ray (feasible solution of affine problem + ray of homogenization). Then you solved modified children to account for the ray so that they generate cuts that will cut out the ray, then you can add these are cuts to the parents and it will hopefully become bounded.

**Comment by odow** [comment]
> I always wondered if anyone had the same need and how it was handled if not like that

We add a big-M a priori bound to the θ variable so that the parent problem is bounded :smile:

**Comment by odow** [comment]
Add section on querying the Jacobian to docs. See
https://discourse.julialang.org/t/struggle-at-using-jump-mathoptinterface-for-calculating-number-of-constraints/116275/2

**Comment by odow** [comment]
Add an example of using DifferentiationInterface to https://jump.dev/JuMP.jl/stable/tutorials/nonlinear/operator_ad/: https://github.com/jump-dev/JuMP.jl/issues/3806

**Comment by odow** [comment]
Here is yet another nested summation case: https://discourse.julialang.org/t/jump-takes-long-time-to-load-the-problem-model/118593

I need to write the canonical explanation of this and the suggested approaches.

**Comment by odow** [comment]
Use or mention PiecewiseAffineApprox.jl in the piecewise linear tutorial: https://jump.dev/JuMP.jl/stable/tutorials/linear/piecewise_linear/

**Comment by odow** [comment]
I think we need a numerical issues tutorial. I link to the Gurobi ones all the time. People are constantly confused about tolerances, etc.

**Comment by jd-foster** [comment]
Classic reference is [Aspects of mathematical modelling related to optimization](https://doi.org/10.1016/0307-904X(81)90002-0)

As of posting, currently active thread on Julia [Discourse](https://discourse.julialang.org/t/history-of-isapprox-in-languages/119347/28) highlights the issues and confusions.

**Comment by odow** [comment]
We now have https://jump.dev/JuMP.jl/dev/tutorials/getting_started/tolerances/, which I quite like. We probably need a separate tutorial to talk about problem scaling, reasonable coefficients, etc. Or perhaps it can be tagged onto the end of this tutorial.

**Comment by odow** [comment]
Somewhat co-incidentally, a few days after I wrote https://jump.dev/JuMP.jl/dev/tutorials/getting_started/tolerances/, `@ambros-gleixner` gave a very similar talk: https://www.youtube.com/watch?v=rKcdF4Fgl-g

We obviously both pilfered from the Gurobi docs, because our examples are variations on a theme!

**Comment by odow** [comment]
Add a section on anti-patterns

**Comment by odow** [comment]
Better document building in parallel: https://github.com/jump-dev/JuMP.jl/issues/3894

**Comment by jd-foster** [comment]
More visual content in the docs would be welcome, I think.
This is a good example for flows in graphs: https://doc.sagemath.org/html/en/thematic_tutorials/linear_programming.html#flows

**Comment by jd-foster** [comment]
Another interesting idea from Sage docs: a compilation page "presenting commands roughly in the order a student would learn the corresponding mathematics" (but for LP not group theory, of course)
https://doc.sagemath.org/html/en/thematic_tutorials/group_theory.html

**Comment by joaquimg** [comment]
Some of these deserve a place in the docs: https://github.com/NREL-Sienna/PowerSimulations.jl/pull/1218#issue-2794848308, possibly at performance tips

**Comment by odow** [comment]
Once registered add a tutorial with https://github.com/samuelsonric/MathOptChordalDecomposition.jl

**Comment by blegat** [comment]
As suggested by @langestefan [on the Julia Slack](https://julialang.slack.com/archives/C681S52FQ/p1747393456926289?thread_ts=1747389389.881179&cid=C681S52FQ), we could clarify how that to create fancy `AbstractArray` subtypes likes the ones of LinearAlgebra, SparseArrays, BlockDiagonal, etc..., you can just create classical JuMP variables and use the normal constructor for these fancy arrays, e.g.,
```julia-repl
julia> using JuMP

julia> model = Model();

julia> @variable(model, x[1:3]);

julia> using SparseArrays

julia> A = sparse([1, 2, 3], [2, 1, 3], x)
3×3 SparseMatrixCSC{VariableRef, Int64} with 3 stored entries:
 ⋅     x[1]  ⋅
 x[2]  ⋅     ⋅
 ⋅     ⋅     x[3]
```

**Comment by odow** [comment]
Document how to drop down to MOI for some features: https://github.com/jump-dev/JuMP.jl/issues/4030

**Comment by odow** [comment]
Document the threading closure bug: https://discourse.julialang.org/t/julia-crashes-without-reporting-anything-when-i-optimize-a-vector-of-models-in-parallel/131801/11?u=odow

**Comment by odow** [comment]
We should add tutorials and documentation for JuliaC/PackageCompiler. We may need to have a separate repo with tests so that they can be added to `solver-tests.yml` and `extension-tests.yml`.

**Comment by franckgaga** [comment]
Hello @odow!

I think you should update the [User-defined operators with vector output](https://jump.dev/JuMP.jl/stable/tutorials/nonlinear/tips_and_tricks/#User-defined-operators-with-vector-outputs) to mention the existence of `MOI.VectorNonlinearOracle` as an alternative, if the operator is used in a constraint.

Otherwise, you could mention the existence of the set in the first warning admonition of <https://jump.dev/JuMP.jl/stable/manual/nonlinear/#jump_user_defined_operators>.

**Comment by odow** [comment]
So my conclusion from mucking about with PackageCompiler and JuliaC for the afternoon is that PackageCompiler just works. I'm hesitant to say we even need a tutorial. 

You just need a package with a `julia_main` entry point:

`MyApp/src/MyApp.jl`
```julia
module MyApp
function julia_main()::Cint
    return 0
end
end
```

Then
```julia
import PackageCompiler
PackageCompiler.create_app("MyApp", "MyAppCompiled")
```

JuliaC, on the other hand, 

```Julia
Verifier error #139: unresolved call from statement (MathOptInterface.set)(JuMP.getfield(model::JuMP.Model, :moi_backend)::MathOptInterface.ModelLike, attr::MathOptInterface.ConstraintName, Base.getfield(cr::JuMP.ConstraintRef{JuMP.Model, MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64}, MathOptInterface.LessThan{Float64}}, JuMP.ScalarShape}, :index)::MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64}, MathOptInterface.LessThan{Float64}}, value::String)::Any
Stacktrace:
 [1] set(model::JuMP.Model, attr::MathOptInterface.ConstraintName, cr::JuMP.ConstraintRef{JuMP.Model, MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64}, MathOptInterface.LessThan{Float64}}, JuMP.ScalarShape}, value::String)
   @ JuMP ~/.julia/packages/JuMP/e83v9/src/optimizer_interface.jl:1294
 [2] set_name(con_ref::JuMP.ConstraintRef{JuMP.Model, MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64}, MathOptInterface.LessThan{Float64}}, JuMP.ScalarShape}, s::String)
   @ JuMP ~/.julia/packages/JuMP/e83v9/src/constraints.jl:423 [inlined]
 [3] add_constraint(model::JuMP.Model, con::JuMP.ScalarConstraint{JuMP.AffExpr, MathOptInterface.LessThan{Float64}}, name::String)
   @ JuMP ~/.julia/packages/JuMP/e83v9/src/constraints.jl:1053
 [4] macro expansion
   @ ~/.julia/packages/JuMP/e83v9/src/macros/@constraint.jl:173 [inlined]
 [5] macro expansion
   @ ~/.julia/packages/JuMP/e83v9/src/macros.jl:402 [inlined]
 [6] run_highs_example()
   @ MyCApp ~/git/jump-dev/ExampleJuMPProject/MyCApp/src/MyCApp.jl:35
 [7] main(ARGS::Vector{String})
   @ MyCApp ~/git/jump-dev/ExampleJuMPProject/MyCApp/src/MyCApp.jl:23
 [8] _main(argc::Int32, argv::Ptr{Ptr{Int8}})
   @ Main ~/.julia/packages/JuliaC/EFQEY/src/scripts/juliac-buildscript.jl:77

Verifier error #140: unresolved call from statement (MathOptInterface.set)(model::HiGHS.Optimizer, %new()::MathOptInterface.RawOptimizerAttribute, φ ()::Any)::Any
Stacktrace:
 [1] empty!(model::HiGHS.Optimizer)
   @ HiGHS ~/.julia/packages/HiGHS/yay5V/src/MOI_wrapper.jl:371
 [2] HiGHS.Optimizer()
   @ HiGHS ~/.julia/packages/HiGHS/yay5V/src/MOI_wrapper.jl:324
 [3] run_highs_example()
   @ MyCApp ~/git/jump-dev/ExampleJuMPProject/MyCApp/src/MyCApp.jl:32
 [4] main(ARGS::Vector{String})
   @ MyCApp ~/git/jump-dev/ExampleJuMPProject/MyCApp/src/MyCApp.jl:23
 [5] _main(argc::Int32, argv::Ptr{Ptr{Int8}})
   @ Main ~/.julia/packages/JuliaC/EFQEY/src/scripts/juliac-buildscript.jl:77
```

Has many many type stability issues all over the place.

**Comment by blegat** [comment]
Yes, even in direct mode, the fact that the type of the moi backend is non-concrete, which is needed for `set_optimizer`, makes it not possible to use with JuliaC. A possible solution would be to have a struct
```julia
struct ConcreteModel{T,M}
    moi_backend::M
    #...
end
const GenericModel{T} = ConcreteModel{T,MOI.ModelLike}
const Model = GenericModel{Float64}
```
Then, we could add a keyword argument like `cement_backend` in `direct_model` and `Model`. When set to `true`, it makes the backend unremoveable. But it would make JuliaC closer to possible, and it could also give small performance gains, since we avoid the type instabilities everytime we access the backend (so we could recommend it in the docs next to `set_string_names_on_creation`).

**Comment by odow** [comment]
@Transurgeon suggested that we could use `distance_to_set` more liberally in the tutorials to convince people they may want to look at the quality of the returned solutions.
